sentry/src/main/java/io/sentry/SentryClient.java

```diff
@@ -171,10 +172,17 @@ private boolean shouldApplyScopeData(
     }
 
     try {
-      final TraceContext traceContext =
-          scope != null && scope.getTransaction() != null
-              ? scope.getTransaction().traceContext()
-              : null;
+      @Nullable TraceContext traceContext = null;
+      if (scope != null) {
+        if (scope.getTransaction() != null) {
+          traceContext = scope.getTransaction().traceContext();
+        } else {
+          final @NotNull PropagationContext propagationContext =
+              TracingUtils.maybeUpdateBaggage(scope, options);
+          traceContext = propagationContext.traceContext();
+        }
+      }
```

Comment:
M: This is probably racey since the lock is just within the `withTraceContext` but you require the `scope.getTransaction()` and `scope.getTransaction().traceContext()`.
Ideally, this whole operation is done within the lock so you guarantee that the value has not been mutated by any other thread.

Reply:
The side effect is that you'd be using the wrong/dirty trace in case that happens.

Reply:
@marandaneto is it possible you're mixing up trace context and propagation context here? The newly added `withPropagationContext` is not used for the first branch of the `if`. And `scope.getTransaction().traceContext()` eventually calls into `updateBaggageValues` on `SentryTracer` which is synchronized.

The `else` branch does utilize the newly added locking as it maybe updates `PropagationContext` and uses the return value of `withPropagationContext` passing it back all the way to here.

Reply:
yes, that was my bad, good point.

I'd thought change it to (L177) -> `transaction = scope.getTransaction()` and reuse the `transaction` object, if another thread modifies the `transaction` between the L177 and 178, you could have an NPE or take the wrong object.

Reply:
Good point, blindly took the existing code without thinking it through. Updated.